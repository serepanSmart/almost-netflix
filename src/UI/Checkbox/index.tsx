import React from 'react';
import {
  CheckboxContainer,
  Icon,
  HiddenCheckbox,
  StyledCheckbox,
} from './styles';

// eslint-disable-next-line max-len
interface CheckboxProps
  extends Omit<
  React.HTMLProps<HTMLInputElement>,
  'ref' | 'as' | 'type' | 'css'
  > {
  square?: boolean;
}

const Checkbox: React.FC<CheckboxProps> = ({
  className,
  checked,
  disabled,
  ...props
}) => (
  <CheckboxContainer className={className}>
    <HiddenCheckbox
      type="checkbox"
      checked={checked}
      disabled={disabled}
      {...props}
    />
    <StyledCheckbox
      data-test-checked={checked}
      checked={checked || false}
      disabled={disabled}
    >
      <Icon viewBox="0 0 24 24">
        <polyline points="20 6 9 17 4 12" />
      </Icon>
    </StyledCheckbox>
  </CheckboxContainer>
);

export default Checkbox;
export type { CheckboxProps };
