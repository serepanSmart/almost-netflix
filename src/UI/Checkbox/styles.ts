import styled from 'styled-components';
import Colors from '../Theme/Colors';

export const CheckboxContainer = styled.div`
  display: inline-flex;
`;

export const Icon = styled.svg`
  polyline {
    fill: none;
    stroke: ${Colors.Scarlet};
    stroke-width: 4px;
  }
`;

export const HiddenCheckbox = styled.input`
  border: 0;
  clip: rect(0 0 0 0);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
`;

export const StyledCheckbox = styled.div<{
  checked: boolean; disabled?: boolean;
}>`
  user-select: none;
  display: inline-block;
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  width: 16px;
  height: 16px;
  background: ${({ disabled }) => (disabled ? Colors.Tide : Colors.White)};
  border: 1px solid ${Colors.Scarlet};
  border-radius: 4px;
  transition: all 100ms;
  border-color: ${({ disabled }) => disabled && Colors.Scarlet} !important;
  &:hover {
    border-color: ${({ disabled }) => (!disabled && Colors.Tide)};
  }
  ${Icon} {
    visibility: ${({ checked }) => (checked ? 'visible' : 'hidden')};
  }
`;