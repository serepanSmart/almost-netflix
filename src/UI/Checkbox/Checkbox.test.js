import Checkbox from '.';
import { render, fireEvent } from '@testing-library/react';

const handleChange = jest.fn();

describe('Checkbox component', () => {
  let component;
  beforeEach(() => {
    component = (props) =>
      render(
        <Checkbox
          checked={false}
          disabled={false}
          onChange={handleChange}
          {...props}
        />,
      );
  });

  describe('Checkbox renders', () => {
    test('should define Checkbox component', () => {
      expect(component()).toBeDefined();
    });

    test('should match snapshot with default props', () => {
      expect(component()).toMatchSnapshot();
    });

    test('should be disabled with "disabled" atrr', () => {
      const { getByRole } = component({ disabled: true });
      expect(getByRole('checkbox')).toBeDisabled();
    });

    test('should be checked with "checked" atrr', () => {
      const { getByRole } = component({ checked: true });
      expect(getByRole('checkbox')).toBeChecked();
    });
  });

  describe('Checkbox events', () => {
    test('should call onChange method', () => {
      const { getByRole } = component();
      const checkbox = getByRole('checkbox');
      fireEvent.click(checkbox);
      expect(handleChange).toHaveBeenCalledTimes(1);
    });
    test('should be checked onChange', () => {
      const { getByRole } = component();
      const checkbox = getByRole('checkbox');
      fireEvent.change(checkbox, {target: {checked: true}});
      expect(checkbox).toBeChecked();
    });
  });
});
