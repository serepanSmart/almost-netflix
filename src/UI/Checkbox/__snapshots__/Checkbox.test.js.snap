// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component Checkbox renders should match snapshot with default props 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c4 polyline {
  fill: none;
  stroke: #f65261;
  stroke-width: 4px;
}

.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline-block;
  cursor: pointer;
  width: 16px;
  height: 16px;
  background: #fff;
  border: 1px solid #f65261;
  border-radius: 4px;
  -webkit-transition: all 100ms;
  transition: all 100ms;
  border-color: !important;
}

.c2:hover {
  border-color: #555555;
}

.c2 .c3 {
  visibility: hidden;
}

<body>
    <div>
      <div
        class="c0"
      >
        <input
          class="c1"
          type="checkbox"
        />
        <div
          class="c2"
          data-test-checked="false"
        >
          <svg
            class="c3 c4"
            viewBox="0 0 24 24"
          >
            <polyline
              points="20 6 9 17 4 12"
            />
          </svg>
        </div>
      </div>
    </div>
  </body>,
  "container": .c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c4 polyline {
  fill: none;
  stroke: #f65261;
  stroke-width: 4px;
}

.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline-block;
  cursor: pointer;
  width: 16px;
  height: 16px;
  background: #fff;
  border: 1px solid #f65261;
  border-radius: 4px;
  -webkit-transition: all 100ms;
  transition: all 100ms;
  border-color: !important;
}

.c2:hover {
  border-color: #555555;
}

.c2 .c3 {
  visibility: hidden;
}

<div>
    <div
      class="c0"
    >
      <input
        class="c1"
        type="checkbox"
      />
      <div
        class="c2"
        data-test-checked="false"
      >
        <svg
          class="c3 c4"
          viewBox="0 0 24 24"
        >
          <polyline
            points="20 6 9 17 4 12"
          />
        </svg>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
